File: samples/_shared/android/SampleTestRunner.kt
// Copyright 2026 Ivan Aliaga <ivan.aliaga@urp.edu.pe>
//
// Licensed under the Apache License, Version 2.0 (the "License");
// you may not use this file except in compliance with the License.
//
// ────────────────────────────────────────────────────────────────────
//  DEV SMOKE HARNESS — this is NOT the product.
// ────────────────────────────────────────────────────────────────────
//
// Fast (<10 s/sample) cross-platform CI verification. Wires a
// `FakeLLMClient` with scripted replies so the pass/fail doesn't
// depend on an on-device GGUF being loaded — only on the wiring:
// Activity boots, Dazzle starts, corpus loads, tool loop fires,
// agent persists messages, report JSON emits. Runs behind the
// `SAMPLE_TEST=1` intent extra; the normal app launch (tap the app
// icon) skips this entirely and uses the real on-device LLM from
// `LLMAdapter.kt` — that path is the actual product.
//
// The on-screen banner stamped by the SampleTestBanner composable
// makes the distinction explicit so nobody confuses the smoke run
// with the production chat.
//
// Matches samples/_shared/ios/SampleTestRunner.swift one-to-one so
// the cross-platform test produces the same report shape.

package dev.dazzle.samples.shared

import android.content.Context
import android.os.Environment
import dev.dazzle.sdk.Agent
import dev.dazzle.sdk.AgentStatus
import dev.dazzle.sdk.Completion
import dev.dazzle.sdk.DazzleServer
import dev.dazzle.sdk.FakeLLMClient
import dev.dazzle.sdk.LLMClient
import dev.dazzle.sdk.Role
import kotlinx.coroutines.delay
import java.io.File

data class SampleTestConfig(
    val sampleName: String,
    /** Pre-scripted assistant completions the FakeLLMClient replays
     *  in order. See `FakeLLMClient`'s contract. */
    val llmScript: List<Completion>,
    /** The question(s) the harness types in on the user's behalf. */
    val userInputs: List<String>,
    val prepare: suspend () -> Unit,
    /** Build the ChatAgent — receives the FakeLLMClient the harness
     *  constructs; the closure wires the same tools + system prompt
     *  the production app uses. */
    val buildAgent: suspend (LLMClient) -> Agent,

    val onAgentReady: ((Agent) -> Unit)? = null,
    val onStatusChange: ((phase: String, detail: String?) -> Unit)? = null,

    val delayBetweenTurnsMs: Long = 1_200L,
    val postRunDisplayMs: Long = 5_000L,
)

data class TestReport(
    val sampleName: String,
    val elapsedMs: Long,
    val turnCount: Int,
    val userTurns: Int,
    val assistantTurns: Int,
    val toolTurns: Int,
    val llmCallCount: Int,
    val lastAssistantText: String,
    val lastToolText: String,
    val status: String,
    val error: String? = null,
)

/**
 * Drives the scripted flow and writes the JSON report. Call from the
 * Activity's `onCreate` (not a Composable) so the lifecycle stays
 * predictable.
 */
suspend fun runSampleTest(context: Context, config: SampleTestConfig): Boolean {
    val start = System.currentTimeMillis()
    val fake = FakeLLMClient(script = config.llmScript)
    var agent: Agent? = null

    return try {
        config.onStatusChange?.invoke("preparing", null)

        if (!DazzleServer.isRunning()) {
            DazzleServer.start(context, DazzleServer.config)
        }

        config.prepare()
        val realMarker = java.io.File("/data/local/tmp/dazzle_real_llm")
        val llm: LLMClient = if (realMarker.exists()) {
            android.util.Log.i("SampleTestRunner",
                "real-LLM marker present → using LLMAdapter")
            try {
                LLMAdapter.makeLLMClient(context)
            } catch (t: Throwable) {
                android.util.Log.e("SampleTestRunner",
                    "LLMAdapter.makeLLMClient failed: ${t.message}", t)
                throw t
            }
        } else {
            fake
        }
        agent = config.buildAgent(llm)
        config.onAgentReady?.invoke(agent)

        delay(400)
        config.onStatusChange?.invoke("running", null)

        // Real LLMs (esp. cloud routers) can take 10–60 s/turn for
        // multi-token replies. Fake is instant; either way 90 s
        // covers it without hanging CI when the network drops.
        val perTurnTimeoutMs = if (realMarker.exists()) 90_000L else 30_000L
        for ((i, input) in config.userInputs.withIndex()) {
            if (i > 0) delay(config.delayBetweenTurnsMs)
            agent.send(input)
            val deadline = System.currentTimeMillis() + perTurnTimeoutMs
            while (agent.status.value != AgentStatus.Idle) {
                if (System.currentTimeMillis() > deadline) {
                    throw RuntimeException(
                        "turn '$input' timed out after " +
                        "${perTurnTimeoutMs / 1000} s")
                }
                delay(100)
            }
        }

        val messages = agent.messages.value
        val assistantTurns = messages.filter { it.role == Role.assistant }
        val toolTurns = messages.filter { it.role == Role.tool }

        val elapsed = System.currentTimeMillis() - start
        config.onStatusChange?.invoke(
            "completed", "${messages.size} turns · $elapsed ms"
        )

        delay(config.postRunDisplayMs)

        writeReport(context, TestReport(
            sampleName = config.sampleName,
            elapsedMs  = System.currentTimeMillis() - start,
            turnCount  = messages.size,
            userTurns  = messages.count { it.role == Role.user },
            assistantTurns = assistantTurns.size,
            toolTurns  = toolTurns.size,
            llmCallCount = fake.callCount,
            lastAssistantText = assistantTurns.lastOrNull()?.text ?: "",
            lastToolText      = toolTurns.lastOrNull()?.text ?: "",
            status = "pass",
            error  = null,
        ))
        writeMarker(context, ok = true, message = "sample_test_${config.sampleName}")
        true
    } catch (t: Throwable) {
        config.onStatusChange?.invoke("failed",
            "${t::class.simpleName}: ${t.message}")
        delay(2_000)
        writeReport(context, TestReport(
            sampleName = config.sampleName,
            elapsedMs  = System.currentTimeMillis() - start,
            turnCount = 0, userTurns = 0, assistantTurns = 0, toolTurns = 0,
            llmCallCount = 0,
            lastAssistantText = "", lastToolText = "",
            status = "fail",
            error  = "${t::class.simpleName}: ${t.message}",
        ))
        writeMarker(context, ok = false, message = "sample_test_${config.sampleName}")
        false
    }
}

/**
 * Convenience: check the intent extras / system property for the
 * SAMPLE_TEST flag. Activities read this from
 * `intent.getStringExtra("SAMPLE_TEST")`.
 */
fun isSampleTestMode(intentExtra: String?): Boolean {
    return intentExtra == "1"
        || System.getProperty("dazzle.sample_test") == "1"
}

// ─────────────────────────────────────────────────────────────────
// Storage note (Android version matrix):
//
//   API 21–29 (Android 5–10):
//       /sdcard/Documents works with WRITE_EXTERNAL_STORAGE +
//       requestLegacyExternalStorage=true in the manifest.
//
//   API 30+ (Android 11+):
//       Scoped storage — /sdcard/Documents blocked for File I/O
//       unless the app holds MANAGE_EXTERNAL_STORAGE (Google Play
//       will REJECT most apps that request it). MediaStore is the
//       sanctioned path but overkill for a dev harness.
//
//   API 33+ (Android 13+):
//       Further tightened. `File.writeText("/sdcard/Documents/…")`
//       fails with EACCES even when the file already exists,
//       depending on the owning UID.
//
//   app-private `context.filesDir` (what we use):
//       Works IDENTICALLY on every Android version from API 21 to
//       35+. No permissions. No MediaStore. The harness shell reads
//       it back via `adb shell run-as <pkg> cat files/…`, which is
//       guaranteed to work on debug-signed builds.
//
// Devs who ship a real app and want user-visible files should use
// getExternalFilesDir() (no permissions, cleared on uninstall) or
// MediaStore Downloads. NEVER /sdcard/Documents direct.
// ─────────────────────────────────────────────────────────────────
private fun writeReport(context: Context, report: TestReport) {
    val file = File(context.filesDir, "sample_test_${report.sampleName}.json")
    // Hand-rolled JSON to avoid pulling kotlinx.serialization into samples
    // that don't already need it (chat-memory has no bundled datasets).
    file.writeText(report.toJson())
}

private fun TestReport.toJson(): String {
    val sb = StringBuilder()
    sb.append("{\n")
    sb.append("  \"sample_name\": \"${sampleName.jsonEscape()}\",\n")
    sb.append("  \"elapsed_ms\": $elapsedMs,\n")
    sb.append("  \"turn_count\": $turnCount,\n")
    sb.append("  \"user_turns\": $userTurns,\n")
    sb.append("  \"assistant_turns\": $assistantTurns,\n")
    sb.append("  \"tool_turns\": $toolTurns,\n")
    sb.append("  \"llm_call_count\": $llmCallCount,\n")
    sb.append("  \"last_assistant_text\": \"${lastAssistantText.jsonEscape()}\",\n")
    sb.append("  \"last_tool_text\": \"${lastToolText.jsonEscape()}\",\n")
    sb.append("  \"status\": \"${status.jsonEscape()}\",\n")
    sb.append("  \"error\": ")
    if (error == null) sb.append("null\n") else sb.append("\"${error.jsonEscape()}\"\n")
    sb.append("}\n")
    return sb.toString()
}

private fun String.jsonEscape(): String = this
    .replace("\\", "\\\\")
    .replace("\"", "\\\"")
    .replace("\n", "\\n")
    .replace("\r", "\\r")
    .replace("\t", "\\t")

private fun writeMarker(context: Context, ok: Boolean, message: String) {
    // filesDir (app-private) — see writeReport for rationale.
    val file = File(context.filesDir, "experiment_backends_complete.marker")
    file.writeText("${System.currentTimeMillis()} " +
                    "${if (ok) "ok" else "error"} $message\n")
}

File: samples/_shared/flutter/lib/src/sample_test_runner.dart
// Copyright 2026 Ivan Aliaga <ivan.aliaga@urp.edu.pe>
// SPDX-License-Identifier: Apache-2.0
//
// Headless end-to-end harness. Runs the SAME production stack — real
// `LLMClient` via `makeLLMClient()` (Qwen 2.5 1.5B GGUF by default),
// real tools, real Dazzle — against a pre-scripted user question, then
// writes a JSON report to the app's Documents dir and exits.
//
// Previously this harness wired a `FakeLLMClient` with hand-written
// replies. That's gone. If the on-device GGUF isn't present, the test
// fails loudly so the operator knows to push it.

import 'dart:convert';
import 'dart:io';

import 'package:dazzle_flutter/dazzle_flutter.dart';
import 'package:path_provider/path_provider.dart';

import 'llm_adapter.dart' show makeLLMClient;

class SampleTestConfig {
  SampleTestConfig({
    required this.sampleName,
    required this.llmScript,
    required this.userInputs,
    required this.prepare,
    required this.buildAgent,
    this.onAgentReady,
    this.onStatusChange,
    this.delayBetweenTurns  = const Duration(milliseconds: 1200),
    this.postRunDisplay     = const Duration(seconds: 5),
  });

  final String sampleName;

  /// Pre-scripted assistant completions the `FakeLLMClient` replays
  /// in order during the dev-smoke run. The real product path (tap
  /// the app icon — no SAMPLE_TEST flag) uses `makeLLMClient()` from
  /// `llm_adapter.dart` and a genuine on-device LLM; this is the
  /// smoke harness, wiring verification only.
  final List<Completion> llmScript;

  final List<String> userInputs;
  final Future<void> Function() prepare;

  final Future<ChatAgent> Function(LLMClient llm) buildAgent;

  final void Function(ChatAgent agent)? onAgentReady;
  final void Function(String phase, String? detail)? onStatusChange;

  final Duration delayBetweenTurns;
  final Duration postRunDisplay;
}

class SampleTestReport {
  SampleTestReport({
    required this.sampleName,
    required this.elapsedMs,
    required this.turnCount,
    required this.userTurns,
    required this.assistantTurns,
    required this.toolTurns,
    required this.llmCallCount,
    required this.lastAssistantText,
    required this.lastToolText,
    required this.status,
    this.error,
  });

  final String sampleName;
  final int elapsedMs;
  final int turnCount;
  final int userTurns;
  final int assistantTurns;
  final int toolTurns;
  final int llmCallCount;
  final String lastAssistantText;
  final String lastToolText;
  final String status;
  final String? error;

  Map<String, Object?> toJson() => {
        'sample_name': sampleName,
        'elapsed_ms': elapsedMs,
        'turn_count': turnCount,
        'user_turns': userTurns,
        'assistant_turns': assistantTurns,
        'tool_turns': toolTurns,
        'llm_call_count': llmCallCount,
        'last_assistant_text': lastAssistantText,
        'last_tool_text': lastToolText,
        'status': status,
        'error': error,
      };
}

/// True when the sample should boot into test mode. Set via
/// `--dart-define=SAMPLE_TEST=1` or by the e2e script setting
/// DAZZLE_SAMPLE_TEST=1 in the env.
bool isSampleTestMode() {
  const fromDefine = String.fromEnvironment('SAMPLE_TEST');
  if (fromDefine == '1') return true;
  if (Platform.environment['DAZZLE_SAMPLE_TEST'] == '1') return true;
  return false;
}

/// Drive the scripted flow and write the JSON report. Returns whether
/// the test passed; samples should exit the process with a non-zero
/// code on failure so the shell runner can detect it.
Future<bool> runSampleTest(SampleTestConfig config) async {
  final stopwatch = Stopwatch()..start();
  final fake = FakeLLMClient(script: config.llmScript);
  ChatAgent? agent;

  // Real LLM path when DAZZLE_REAL_LLM=1 is set via any of:
  //   * `--dart-define=DAZZLE_REAL_LLM=1` at build time (works on
  //     every platform; the pragma compiles into the kernel),
  //   * `Platform.environment['DAZZLE_REAL_LLM']` (the iOS sim
  //     simctl-launch path: SIMCTL_CHILD_DAZZLE_REAL_LLM=1),
  //   * a `/data/local/tmp/dazzle_real_llm` marker file (Android
  //     intent extras don't reach Platform.environment, so the
  //     marker is the easier signal from `adb shell` invocations).
  const _realLlmDefine = String.fromEnvironment('DAZZLE_REAL_LLM');
  bool useRealLlm = _realLlmDefine == '1' ||
                    Platform.environment['DAZZLE_REAL_LLM'] == '1';
  if (!useRealLlm) {
    try {
      useRealLlm = await File('/data/local/tmp/dazzle_real_llm').exists();
    } catch (_) {/* iOS / desktop — file path doesn't exist, ignore */}
  }

  try {
    config.onStatusChange?.call('preparing', null);
    await DazzleServer.shared.waitForReady();

    await config.prepare();
    final LLMClient llmForAgent;
    if (useRealLlm) {
      // ignore: avoid_print
      print('[sample_test] DAZZLE_REAL_LLM=1 → using real LLMAdapter');
      llmForAgent = await makeLLMClient();
    } else {
      llmForAgent = fake;
    }
    agent = await config.buildAgent(llmForAgent);
    config.onAgentReady?.call(agent);

    // Tiny breather so the outer app has a frame to mount the ChatScreen
    // bound to this agent before messages start streaming in.
    await Future<void>.delayed(const Duration(milliseconds: 400));
    config.onStatusChange?.call('running', null);

    final perTurnTimeout = useRealLlm
        ? const Duration(seconds: 90)
        : const Duration(seconds: 30);
    for (var i = 0; i < config.userInputs.length; i++) {
      final input = config.userInputs[i];
      if (i > 0) {
        // Inter-turn pause so a viewer can read the previous
        // assistant reply before the next user message lands.
        await Future<void>.delayed(config.delayBetweenTurns);
      }
      await agent.send(input);
      final deadline = DateTime.now().add(perTurnTimeout);
      while (agent.status.value != AgentStatus.idle) {
        if (DateTime.now().isAfter(deadline)) {
          throw StateError("turn '$input' timed out after ${perTurnTimeout.inSeconds} s");
        }
        await Future<void>.delayed(const Duration(milliseconds: 100));
      }
    }

    final messages = agent.messages.value;
    final assistantTurns =
        messages.where((t) => t.role == Role.assistant).toList();
    final toolTurns = messages.where((t) => t.role == Role.tool).toList();

    config.onStatusChange?.call(
        'completed',
        '${messages.length} turns · ${stopwatch.elapsedMilliseconds} ms');

    // Visible breathing time so the user can read the final reply
    // before the harness yanks the process.
    await Future<void>.delayed(config.postRunDisplay);

    await _writeReport(SampleTestReport(
      sampleName: config.sampleName,
      elapsedMs: stopwatch.elapsedMilliseconds,
      turnCount: messages.length,
      userTurns: messages.where((t) => t.role == Role.user).length,
      assistantTurns: assistantTurns.length,
      toolTurns: toolTurns.length,
      llmCallCount: fake.callCount,
      lastAssistantText:
          assistantTurns.isNotEmpty ? assistantTurns.last.text : '',
      lastToolText: toolTurns.isNotEmpty ? toolTurns.last.text : '',
      status: 'pass',
    ));
    await _writeMarker(
        ok: true, message: 'sample_test_${config.sampleName}');
    return true;
  } catch (e, st) {
    // ignore: avoid_print
    print('sample_test ${config.sampleName} failed: $e\n$st');
    config.onStatusChange?.call('failed', '$e');
    // Hold the error on screen briefly so the viewer can read it.
    await Future<void>.delayed(const Duration(seconds: 2));
    await _writeReport(SampleTestReport(
      sampleName: config.sampleName,
      elapsedMs: stopwatch.elapsedMilliseconds,
      turnCount: 0,
      userTurns: 0,
      assistantTurns: 0,
      toolTurns: 0,
      llmCallCount: 0,
      lastAssistantText: '',
      lastToolText: '',
      status: 'fail',
      error: '$e',
    ));
    await _writeMarker(
        ok: false, message: 'sample_test_${config.sampleName}');
    return false;
  } finally {
    await agent?.close();
    stopwatch.stop();
  }
}

Future<Directory> _reportDir() async {
  // On Android, the public Documents dir isn't writable without extra
  // permissions in Flutter plugins; fall back to the app's Documents
  // dir so the harness always succeeds.
  try {
    final dir = await getApplicationDocumentsDirectory();
    if (!dir.existsSync()) dir.createSync(recursive: true);
    return dir;
  } catch (_) {
    final dir = Directory.systemTemp;
    return dir;
  }
}

Future<void> _writeReport(SampleTestReport report) async {
  final dir = await _reportDir();
  final file = File('${dir.path}/sample_test_${report.sampleName}.json');
  const enc = JsonEncoder.withIndent('  ');
  await file.writeAsString(enc.convert(report.toJson()));
}

Future<void> _writeMarker({required bool ok, required String message}) async {
  final dir = await _reportDir();
  final f = File('${dir.path}/experiment_backends_complete.marker');
  final ms = DateTime.now().millisecondsSinceEpoch;
  await f.writeAsString('$ms ${ok ? "ok" : "error"} $message\n');
}
